src/lib/FreeC/LiftedIR/Converter/Expr.hs

```diff
@@ -332,14 +332,51 @@ rawBind srcSpan mx x varType expr = do
 --   The given expression is the right-hand side of the let.
 liftBinds :: [IR.Bind] -> IR.Expr -> Converter LIR.Expr
 liftBinds [] expr = liftExpr expr
-liftBinds
-  ((IR.Bind srcSpan (IR.VarPat patSrcSpan ident varPatType isStrict) bindExpr)
-   : bs) expr = localEnv $ do
-    _ <- renameAndDefineLIRVar srcSpan isStrict ident varPatType
-    expr' <- liftBinds bs expr
-    patType' <- mapM LIR.liftType varPatType
-    varPat' <- makeVarPat patSrcSpan (IR.UnQual $ IR.Ident ident) patType'
-    shareType' <- mapM LIR.liftType' varPatType
-    bindExpr' <- liftExpr bindExpr
-    let shareExpr = LIR.Share srcSpan bindExpr' shareType'
-    return $ LIR.Bind srcSpan shareExpr (LIR.Lambda srcSpan [varPat'] expr')
+liftBinds ((IR.Bind srcSpan varPat bindExpr) : bs) expr = localEnv $ do
+  let (IR.VarPat patSrcSpan ident varPatType isStrict) = varPat
+  _ <- renameAndDefineLIRVar srcSpan isStrict ident varPatType
+  expr' <- liftBinds bs expr
+  patType' <- mapM LIR.liftType varPatType
+  varPat' <- makeVarPat patSrcSpan (IR.varPatQName varPat) patType'
+  shareType' <- mapM LIR.liftType' varPatType
+  bindExpr' <- liftExpr bindExpr
+  let countExprs = expr : map IR.bindExpr bs
+      shareOp
+        = if sum (map (countVarInExpr $ IR.varPatQName varPat) countExprs) > 1
+          then LIR.Share
+          else LIR.Call
+      shareExpr  = shareOp srcSpan bindExpr' shareType'
+  return $ LIR.Bind srcSpan shareExpr (LIR.Lambda srcSpan [varPat'] expr')
+
+-- | Counts the number of times the variable with the given qualified name
+--   occurs in the given expression.
+countVarInExpr :: IR.QName -> IR.Expr -> Int
+countVarInExpr varPat = countVarInExpr'
+ where
+  countVarInExpr' :: IR.Expr -> Int
+  countVarInExpr' IR.Con {}                    = 0
+  countVarInExpr' (IR.Var _ varName _)
+    = if varPat == varName then 1 else 0
```

Comment:
Consider to rename `varPat` since it is not an `IR.VarPat` but the `IR.VarName` of an `IR.VarPat`.

```suggestion
countVarInExpr :: IR.QName -> IR.Expr -> Int
countVarInExpr varName = countVarInExpr'
 where
  countVarInExpr' :: IR.Expr -> Int
  countVarInExpr' IR.Con {}             = 0
  countVarInExpr' (IR.Var _ varName' _) | varName == varName' = 1
                                        | otherwise = 0
```